File: cpp/src/packedtileid.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Navigation Data Standard e.V. - See "LICENSE" file.

#include "packedtileid.h"

#include "glm/glm.hpp"
#include "glm/ext.hpp"
#include <stdexcept>

namespace ndsmath
{

PackedTileId::PackedTileId() : value_(0u) {}

PackedTileId::PackedTileId(uint32_t value) : value_(value) {}

namespace
{
void validateLevel(int level)
{
    if (level < 0 || level > 15)
    {
        throw std::out_of_range("PackedTileId level must be in [0, 15]");
    }
}

// Deinterleave a tile's Morton number into its (x, y) tile-grid indices.
// Per the NDS Morton scheme, X has (level+1) bits and Y has level bits.
// Mirrors the Python reference's _deinterleave_morton.
void deinterleaveMorton(uint32_t morton, int level, uint32_t &x, uint32_t &y)
{
    x = 0;
    y = 0;
    for (int i = 0; i < level; ++i)
    {
        if (morton & (1u << (2 * i)))
            x |= (1u << i);
        if (morton & (1u << (2 * i + 1)))
            y |= (1u << i);
    }
    if (morton & (1u << (2 * level)))
        x |= (1u << level); // extra X bit
}

// Inverse of deinterleaveMorton.
uint32_t interleaveCoords(uint32_t x, uint32_t y, int level)
{
    uint32_t morton = 0;
    for (int i = 0; i < level; ++i)
    {
        if (x & (1u << i))
            morton |= (1u << (2 * i));
        if (y & (1u << i))
            morton |= (1u << (2 * i + 1));
    }
    if (x & (1u << level))
        morton |= (1u << (2 * level));
    return morton;
}
} // namespace

PackedTileId PackedTileId::fromValue(int32_t value)
{
    PackedTileId tile(static_cast<uint32_t>(value));
    if (!tile.isValid())
    {
        throw std::out_of_range("PackedTileId value is invalid");
    }
    return tile;
}

PackedTileId PackedTileId::fromTileIndex(uint32_t mortonNumber, int level)
{
    validateLevel(level);
    const auto maxMorton = (uint64_t{1} << (2 * level + 1)) - 1;
    if (static_cast<uint64_t>(mortonNumber) > maxMorton)
    {
        throw std::out_of_range("PackedTileId morton number exceeds level range");
    }
    return PackedTileId(mortonNumber + (1u << (16 + level)));
}

PackedTileId PackedTileId::fromTileXY(uint32_t x, uint32_t y, int level)
{
    validateLevel(level);
    const auto maxX = (uint32_t{1} << (level + 1)) - 1;
    const auto maxY = (uint32_t{1} << level) - 1;
    if (x > maxX || y > maxY)
    {
        throw std::out_of_range("PackedTileId tile-grid coordinate exceeds level range");
    }
    return fromTileIndex(interleaveCoords(x, y, level), level);
}

PackedTileId PackedTileId::fromNdsCoordinates(int32_t x, int32_t y, int level)
{
    validateLevel(level);
    return PackedTileId(MortonCode::fromNdsCoordinates(x, y), level);
}

PackedTileId PackedTileId::fromWgs84(double longitude, double latitude, int level)
{
    validateLevel(level);
    int32_t x;
    int32_t y;
    HighPrecWgs84(longitude, latitude).toNdsCoordinates(x, y);
    return fromNdsCoordinates(x, y, level);
}

// Neighbour traversal: deinterleave to (x, y) tile-grid indices, step by one
// with wraparound, and reinterleave. Because the tile numbering follows the
// two's-complement Morton ordering, an unsigned +/-1 with wrap is a correct
// geographic step in every direction, including the antimeridian (X) and
// pole (Y) wraps.

PackedTileId PackedTileId::westNeighbour() const
{
    const int lvl = level();
    uint32_t x, y;
    deinterleaveMorton(mortonNumber(), lvl, x, y);
    const uint32_t maxX = (1u << (lvl + 1)) - 1;
    x = (x - 1) & maxX;
    return fromTileIndex(interleaveCoords(x, y, lvl), lvl);
}

PackedTileId PackedTileId::eastNeighbour() const
{
    const int lvl = level();
    uint32_t x, y;
    deinterleaveMorton(mortonNumber(), lvl, x, y);
    const uint32_t maxX = (1u << (lvl + 1)) - 1;
    x = (x + 1) & maxX;
    return fromTileIndex(interleaveCoords(x, y, lvl), lvl);
}

PackedTileId PackedTileId::southNeighbour() const
{
    const int lvl = level();
    uint32_t x, y;
    deinterleaveMorton(mortonNumber(), lvl, x, y);
    const uint32_t maxY = (1u << lvl) - 1;
    y = (y - 1) & maxY;
    return fromTileIndex(interleaveCoords(x, y, lvl), lvl);
}

PackedTileId PackedTileId::northNeighbour() const
{
    const int lvl = level();
    uint32_t x, y;
    deinterleaveMorton(mortonNumber(), lvl, x, y);
    const uint32_t maxY = (1u << lvl) - 1;
    y = (y + 1) & maxY;
    return fromTileIndex(interleaveCoords(x, y, lvl), lvl);
}

PackedTileId::PackedTileId(MortonCode mortonCode, const int level)
{
    validateLevel(level);

    int32_t xCoord;
    int32_t yCoord;

    mortonCode.toNdsCoordinates(xCoord, yCoord);

    int64_t xCoord2 = xCoord;
    int64_t yCoord2 = yCoord;

    // Wrap negative coordinates into the unsigned range via two's complement
    // (add 2^32 for x, 2^31 for y), matching the reference implementation.
    if (xCoord2 < 0)
    {
        xCoord2 += (1LL << 32);
    }

    if (yCoord2 < 0)
    {
        yCoord2 += (1LL << 31);
    }

    int nLevel = 31 - level;

    int64_t nX = (xCoord2) >> nLevel;
    int64_t nY = (yCoord2) >> nLevel;

    MortonCode temp = MortonCode::fromNdsCoordinates(nX, nY);
    value_ = temp.value();
    value_ += (1u << (16 + level));
}

bool PackedTileId::isValid() const
{
    const auto MIN_PACKED_TILE_ID = 1u << 16u;
    if (value_ < MIN_PACKED_TILE_ID)
    {
        return false;
    }

    const auto tileLevel = level();
    if (tileLevel > 15)
    {
        return false;
    }

    const auto maxMorton = (uint64_t{1} << (2 * tileLevel + 1)) - 1;
    return static_cast<uint64_t>(mortonNumber()) <= maxMorton;
}

MortonCode PackedTileId::southWestCorner() const
{
    const uint64_t tileNum = mortonNumber();
    return MortonCode(tileNum << (63 - (2 * level() + 1)));
}

MortonCode PackedTileId::northEastCorner() const
{
    int32_t x, y;
    southWestCorner().toNdsCoordinates(x, y);
    auto const s = size();
    return MortonCode::fromNdsCoordinates(x + s, y + s);
}

void PackedTileId::center(int32_t &centerX, int32_t &centerY) const
{
    southWestCorner().toNdsCoordinates(centerX, centerY);
    auto const halfSize = size() / 2;
    centerX += halfSize;
    centerY += halfSize;
}

uint32_t PackedTileId::mortonNumber() const
{
    const auto tileLevel = level();
    return (value_ - (1u << (16 + tileLevel)));
}

uint32_t PackedTileId::x() const
{
    uint32_t resultX;
    uint32_t resultY;
    deinterleaveMorton(mortonNumber(), level(), resultX, resultY);
    return resultX;
}

uint32_t PackedTileId::y() const
{
    uint32_t resultX;
    uint32_t resultY;
    deinterleaveMorton(mortonNumber(), level(), resultX, resultY);
    return resultY;
}

uint32_t PackedTileId::size() const
{
    return 1u << (31 - (level()));
}

int PackedTileId::level() const
{
    int level = 0;

    auto tID = value_ >> 16u;
    for (; tID > 1u; level++)
    {
        tID = tID >> 1;
    }

    return level;
}

bool PackedTileId::operator==(const PackedTileId &other) const
{
    return value_ == other.value_;
}

bool PackedTileId::operator!=(const PackedTileId &other) const
{
    return value_ != other.value_;
}

bool PackedTileId::operator<(const PackedTileId &other) const
{
    return value_ < other.value_;
}

int32_t PackedTileId::value() const
{
    // Stored unsigned for clean bit math; the NDS.Live API value is signed
    // int32 (level-15 tiles are negative). Reinterpret the bit pattern.
    return static_cast<int32_t>(value_);
}

PackedTileId::operator uint32_t() const
{
    // Unsigned bit pattern, used for hashing / std::set keys.
    return value_;
}

PackedTileIds getTileIdsForBoundingBox(int32_t swX, int32_t swY, int32_t neX, int32_t neY,
                                       int level)
{
    validateLevel(level);
    PackedTileIds tileIds;

    // Calculate tile size at this level
    const uint32_t tileSize = 1u << (31 - level);

    auto const floorDiv = [](int64_t a, int64_t b)
    {
        auto q = a / b;
        auto r = a % b;
        if (r != 0 && ((a < 0) != (b < 0)))
        {
            --q;
        }
        return q;
    };

    const int64_t startTileX = floorDiv(swX, static_cast<int64_t>(tileSize));
    const int64_t startTileY = floorDiv(swY, static_cast<int64_t>(tileSize));
    const int64_t endTileX = floorDiv(neX, static_cast<int64_t>(tileSize));
    const int64_t endTileY = floorDiv(neY, static_cast<int64_t>(tileSize));

    // Iterate through all tiles in the bounding box
    for (int64_t tileY = startTileY; tileY <= endTileY; ++tileY)
    {
        for (int64_t tileX = startTileX; tileX <= endTileX; ++tileX)
        {
            // Calculate the south-west corner of this tile
            const int64_t tileSwX = tileX * static_cast<int64_t>(tileSize);
            const int64_t tileSwY = tileY * static_cast<int64_t>(tileSize);

            // Create morton code from the tile's south-west corner
            const MortonCode morton = MortonCode::fromNdsCoordinates(static_cast<int32_t>(tileSwX),
                                                                     static_cast<int32_t>(tileSwY));

            // Create the packed tile ID
            const PackedTileId tileId(morton, level);
            tileIds.push_back(tileId);
        }
    }

    return tileIds;
}

} // namespace ndsmath

File: java/src/main/java/io/github/ndsev/ndslivemath/PackedTileId.java
// SPDX-License-Identifier: BSD-3-Clause
package io.github.ndsev.ndslivemath;

/**
 * Represents a tile in the hierarchical NDS.Live tiling system.
 *
 * <p>
 * Per the NDS.Live standard, tile IDs are signed 32-bit integers. For levels
 * 0-14, values are positive. For level 15, values are negative
 * ({@code -2147483648} to {@code -1}) because the level bit (bit 31) is the
 * sign bit in signed int32 representation.
 * </p>
 *
 * <p>
 * Internally values are stored as unsigned 32-bit integers (in a {@code long}
 * masked to {@code 0xFFFFFFFF}) to enable clean bit operations. Conversion
 * between signed and unsigned happens only at the API boundary: the constructor
 * accepts both representations and {@link #value()} returns the signed int32
 * per the standard.
 * </p>
 *
 * <p>
 * Faithful port of {@code python/src/ndslive/math/tileid.py}.
 * </p>
 */
public final class PackedTileId {

	private static final long TWO_POW_32 = 1L << 32;
	private static final long TWO_POW_31 = 1L << 31;

	/** Internal unsigned 32-bit storage, held in a long masked to 0xFFFFFFFF. */
	private final long value;

	/** Construct from tile id value 0 (invalid; provided for symmetry only). */
	// No default constructor: a value of 0 is always invalid, so we omit it.

	/**
	 * Construct a PackedTileId from a tile ID value.
	 *
	 * <p>
	 * Accepts both signed and unsigned 32-bit integer representations. Level 15
	 * tiles have negative signed values ({@code -2147483648} to {@code -1}); levels
	 * 0-14 have positive values.
	 * </p>
	 *
	 * @param value
	 *            tile ID as signed int32 (negative for level 15) or unsigned
	 * @throws IllegalArgumentException
	 *             if the tile ID is invalid
	 */
	public PackedTileId(long value) {
		long v;
		if (value < 0) {
			// Convert signed int32 to unsigned (e.g. -2147483648 -> 2147483648).
			v = value + TWO_POW_32;
		} else if (value >= TWO_POW_32) {
			// Mask values outside 32-bit range to 32 bits.
			v = value & 0xFFFFFFFFL;
		} else {
			v = value;
		}
		this.value = v;
		validate();
	}

	/**
	 * Get the tile ID value as signed int32 per the NDS.Live standard. Level 15
	 * tiles return negative values; levels 0-14 return positive values.
	 *
	 * @return the signed int32 tile id value
	 */
	public int value() {
		if (this.value >= TWO_POW_31) {
			return (int) (this.value - TWO_POW_32);
		}
		return (int) this.value;
	}

	/**
	 * Create a PackedTileId directly from a tile morton number and level, without
	 * any coordinate conversion.
	 *
	 * @param mortonNumber
	 *            the tile's morton number ({@code 0} to {@code 2^(2*level+1) - 1})
	 * @param level
	 *            tile level (0-15)
	 * @return the PackedTileId with the given morton number at the given level
	 * @throws IllegalArgumentException
	 *             if level or mortonNumber are out of range
	 */
	public static PackedTileId fromTileIndex(long mortonNumber, int level) {
		if (level < 0 || level > 15) {
			throw new IllegalArgumentException("Invalid level " + level + " (must be 0-15)");
		}
		long maxMorton = (1L << (2 * level + 1)) - 1;
		if (mortonNumber < 0 || mortonNumber > maxMorton) {
			throw new IllegalArgumentException("Invalid morton number " + mortonNumber + " for level " + level
					+ " (allowed: 0-" + maxMorton + ")");
		}
		long value = mortonNumber + (1L << (16 + level));
		return new PackedTileId(value);
	}

	/**
	 * Create a PackedTileId for the tile at {@code level} that contains the
	 * full-precision NDS coordinates encoded in {@code mortonCode}.
	 *
	 * @param mortonCode
	 *            a MortonCode representing full-precision NDS coordinates
	 * @param level
	 *            tile level (0-15)
	 * @return the PackedTileId of the tile containing the encoded point
	 * @throws IllegalArgumentException
	 *             if level is out of range
	 */
	public static PackedTileId fromMortonAndLevel(MortonCode mortonCode, int level) {
		if (level < 0 || level > 15) {
			throw new IllegalArgumentException("Invalid level " + level + " (must be 0-15)");
		}

		long[] coords = mortonCode.toNdsCoordinates();
		long xCoord = coords[0];
		long yCoord = coords[1];

		if (xCoord < 0) {
			xCoord += TWO_POW_32;
		}
		if (yCoord < 0) {
			yCoord += TWO_POW_31;
		}

		int nLevel = 31 - level;
		long nX = xCoord >>> nLevel;
		long nY = yCoord >>> nLevel;

		MortonCode temp = MortonCode.fromNdsCoordinates(nX, nY);
		long value = temp.value() + (1L << (16 + level));
		return new PackedTileId(value);
	}

	/**
	 * @return the level of the tile (0-15)
	 */
	public int level() {
		int level = 0;
		long tileId = this.value >>> 16;
		while (tileId > 1) {
			tileId >>>= 1;
			level += 1;
		}
		return level;
	}

	/**
	 * @return the size of the tile in NDS coordinate units
	 */
	public long size() {
		return 1L << (31 - level());
	}

	/**
	 * Get tile dimensions in meters at the tile's center latitude.
	 *
	 * @return a {@code double[]} of {@code {width_meters, height_meters}}
	 */
	public double[] dimensionsInMeters() {
		long[] center = center();
		Wgs84 centerWgs = Wgs84.fromNdsCoordinates(center[0], center[1]);
		long tileSize = size();
		return Wgs84.ndsDistanceToMeters(tileSize, tileSize, centerWgs.y);
	}

	/**
	 * @return the center of the tile in NDS coordinates as {@code {x, y}}
	 */
	public long[] center() {
		long[] sw = southWestCorner();
		long halfSize = size() / 2;
		return new long[]{sw[0] + halfSize, sw[1] + halfSize};
	}

	/**
	 * @return the south-west corner of the tile in NDS coordinates as {@code {x,
	 *         y}}
	 */
	public long[] southWestCorner() {
		long mortonNumber = mortonNumber();
		long shifted = mortonNumber << (63 - (2 * level() + 1));
		return new MortonCode(shifted).toNdsCoordinates();
	}

	/**
	 * Returns the north-east corner of the tile in NDS coordinates. This corner is
	 * <em>exclusive</em> (the first point outside the tile).
	 *
	 * @return the north-east corner as {@code {x, y}}
	 */
	public long[] northEastCorner() {
		long[] sw = southWestCorner();
		long size = size();
		return new long[]{sw[0] + size, sw[1] + size};
	}

	/**
	 * @return the Morton number of the tile (the value minus the level offset)
	 */
	public long mortonNumber() {
		int tileLevel = level();
		return this.value - (1L << (16 + tileLevel));
	}

	private void validate() {
		long minPackedTileId = 1L << 16;
		if (this.value < minPackedTileId) {
			throw new IllegalArgumentException("Invalid PackedTileId(" + value() + "): value must be >= "
					+ minPackedTileId + " or negative for level 15");
		}

		int tileLevel = level();
		long morton = mortonNumber();
		long maxMorton = (1L << (2 * tileLevel + 1)) - 1;

		if (morton < 0 || morton > maxMorton) {
			throw new IllegalArgumentException("Invalid PackedTileId(" + value() + "): morton number " + morton
					+ " exceeds valid range for level " + tileLevel + " (allowed: 0-" + maxMorton + ")");
		}
	}

	private static long[] deinterleaveMorton(long morton, int level) {
		long x = 0;
		long y = 0;
		for (int i = 0; i < level; i++) {
			if ((morton & (1L << (2 * i))) != 0) {
				x |= (1L << i);
			}
			if ((morton & (1L << (2 * i + 1))) != 0) {
				y |= (1L << i);
			}
		}
		if ((morton & (1L << (2 * level))) != 0) {
			x |= (1L << level);
		}
		return new long[]{x, y};
	}

	private static long interleaveCoords(long x, long y, int level) {
		long morton = 0;
		for (int i = 0; i < level; i++) {
			if ((x & (1L << i)) != 0) {
				morton |= (1L << (2 * i));
			}
			if ((y & (1L << i)) != 0) {
				morton |= (1L << (2 * i + 1));
			}
		}
		if ((x & (1L << level)) != 0) {
			morton |= (1L << (2 * level));
		}
		return morton;
	}

	/**
	 * @return the tile to the west at the same level, wrapping at the antimeridian
	 */
	public PackedTileId westNeighbour() {
		int level = level();
		long[] xy = deinterleaveMorton(mortonNumber(), level);
		long maxX = (1L << (level + 1)) - 1;
		long x = (xy[0] - 1) & maxX;
		long newMorton = interleaveCoords(x, xy[1], level);
		return fromTileIndex(newMorton, level);
	}

	/**
	 * @return the tile to the east at the same level, wrapping at the antimeridian
	 */
	public PackedTileId eastNeighbour() {
		int level = level();
		long[] xy = deinterleaveMorton(mortonNumber(), level);
		long maxX = (1L << (level + 1)) - 1;
		long x = (xy[0] + 1) & maxX;
		long newMorton = interleaveCoords(x, xy[1], level);
		return fromTileIndex(newMorton, level);
	}

	/**
	 * @return the tile to the south at the same level, wrapping at the south pole
	 */
	public PackedTileId southNeighbour() {
		int level = level();
		long[] xy = deinterleaveMorton(mortonNumber(), level);
		long maxY = (1L << level) - 1;
		long y = (xy[1] - 1) & maxY;
		long newMorton = interleaveCoords(xy[0], y, level);
		return fromTileIndex(newMorton, level);
	}

	/**
	 * @return the tile to the north at the same level, wrapping at the north pole
	 */
	public PackedTileId northNeighbour() {
		int level = level();
		long[] xy = deinterleaveMorton(mortonNumber(), level);
		long maxY = (1L << level) - 1;
		long y = (xy[1] + 1) & maxY;
		long newMorton = interleaveCoords(xy[0], y, level);
		return fromTileIndex(newMorton, level);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (!(obj instanceof PackedTileId)) {
			return false;
		}
		return this.value == ((PackedTileId) obj).value;
	}

	@Override
	public int hashCode() {
		return Long.hashCode(this.value);
	}

	@Override
	public String toString() {
		return "PackedTileId(value=" + value() + ")";
	}
}
